File: src/main/java/org/hisp/dhis/lib/expression/spi/IllegalExpressionException.java
package org.hisp.dhis.lib.expression.spi;

import lombok.Getter;

import java.util.List;

import static java.lang.String.format;

@Getter
public class IllegalExpressionException extends IllegalArgumentException {

    private final transient List<Issue> errors;
    private final transient List<Issue> warnings;

    public IllegalExpressionException(String message) {
        super(message);
        this.errors = List.of(new Issue(() -> "", message));
        this.warnings = List.of();
    }

    public IllegalExpressionException(List<Issue> errors, List<Issue> warnings) {
        super(format("%d error(s), %d warning(s)", errors.size(), warnings.size()));
        this.errors = errors;
        this.warnings = warnings;
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/spi/IllegalExpressionException.kt
package org.hisp.dhis.lib.expression.spi

class IllegalExpressionException : IllegalArgumentException {
    @Transient
    private val errors: List<Issue>

    @Transient
    private val warnings: List<Issue>

    constructor(message: String) : super(message) {
        errors = listOf(Issue({ "" }, message))
        warnings = listOf()
    }

    constructor(errors: List<Issue>, warnings: List<Issue>) : super(
        String.format(
            "%d error(s), %d warning(s)",
            errors.size,
            warnings.size)) {
        this.errors = errors
        this.warnings = warnings
    }

    fun getErrors() : List<Issue> {
        return errors;
    }

    fun getWarnings() : List<Issue> {
        return warnings;
    }
}
